Delay initialization of SocketsHttpHandler specific HTTP2+ data

marek-safar opened the pull request:
It avoids generating data that are never used and also saves about 4k in size for browser

Contributes to #44534

marek-safar committed:
Delay initialization of SocketsHttpHandler specific HTTP2+ data

It saves about 4k for browser

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpMethod.cs:
```diff
@@ -175,13 +169,25 @@ public override string ToString()
+            if (method._http3Index != null)
+                return method;
+
+            return method._method.ToUpperInvariant() switch
```
This potentially allocates on every call.  It won't for the built-in instances, but it very well could for others.  This must not be merged.

marek-safar replied:
I thought it's a very uncommon but good point. Fixed the code

stephentoub commented:
cc: @geoffkizer, @scalablecory

scalablecory commented:
Rather than lazy initializing, can we simply not include these pieces of data if they aren't use?

If not, I'm curious what effect this will have on perf -- suddenly our main header loop will have quite a bit more code in it. Do you have a benchmark for this?

geoffkizer commented:
If the concern is dll size, then as @scalablecory says, let's just remove this stuff if it's not used.

marek-safar committed:
Rewrite string allocation code path

marek-safar committed:
Update HttpMethod.cs

marek-safar committed:
Update HttpMethod.cs

karelz commented:
@marek-safar can you please answer the questions above?
Is it helping Linker effort to remove all HTTP/2 and HTTP/3 and this is rooting it currently?

karelz commented:
@marek-safar ping?

marek-safar commented:
> Rather than lazy initializing, can we simply not include these pieces of data if they aren't use?

That's what this change does. It allows illinker to trim the data and code if they are not needed in the user app. Without this change illinker cannot do that because Http3/Http2 data are always required due to static ctor dependency.

>I'm curious what effect this will have on perf -- suddenly our main header loop will have quite a bit more code in it.

I'm not sure what you mean by that. The code is executed only once as before just not in static ctor. I don't see where any perf changes should appear. There could be a positive perf impact for Trace and Patch methods which were not cached before and are now though.

karelz commented:
@geoffkizer and @scalablecory still think there should be better way to do this with less impact on common path -- @geoffkizer can you please comment more here?

geoffkizer commented:
> I don't see where any perf changes should appear.

There's added code in Http2EncodedName, Http3EncodedName, etc to check if the property is initialized yet. These properties are used when we are formatting requests on the wire.

It seems like we should be able to achieve the same result in some way that doesn't introduce these added checks.

marek-safar committed:
Revert change to KnownHeader

marek-safar committed:
Merge remote-tracking branch 'upstream/master' into http-cleanup

marek-safar committed:
Merge branch 'http-cleanup' of github.com:marek-safar/runtime into http-cleanup

marek-safar commented:
Removed the code which added the ifs (though I don't think it mattered when the caller is 100 times more expensive) to unlock the merge. I also updated the issue to track the undone work.

geoffkizer commented:
The deferred method dictionary initialization seems fine to me.

@scalablecory are you ok with the HTTP3 changes here?

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpMethod.cs:
```diff
@@ -175,13 +171,29 @@ public override string ToString()
+            if (method._http3Index != null)
+                return method;
+
+            if (s_knownMethods is null)
+            {
+                s_knownMethods = new Dictionary<HttpMethod, HttpMethod>(9)
```
Technically this should use a Volatile.Read/Write:
```C#
Dictionary<HttpMethod, HttpMethod>? knownMethods = Volatile.Read(ref s_knownMethods);
if (knownMethods is null)
{
    knownMethods = new Dictionary<...>(...) { ... };
    Volatile.Write(ref s_knownMethods, knownMethods);
}

return knownMethods.TryGetValue(...
```

geoffkizer replied:
Given how small this dictionary is, I wonder if we shouldn't just write some hard-wired code to check for the specific methods. That would mean less memory usage (no dictionary) and would probably be faster as well.

stephentoub replied:
Yeah, we should:
https://github.com/dotnet/runtime/pull/47281

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpMethod.cs:
```diff
@@ -89,9 +80,16 @@ public string Method
+
+                return _http3EncodedBytes = _http3Index is int index && index >= 0 ?
+                    QPackEncoder.EncodeStaticIndexedHeaderFieldToArray(index) :
+                    QPackEncoder.EncodeLiteralHeaderFieldWithStaticNameReferenceToArray(H3StaticTable.MethodGet, _method);
```
This similarly doesn't look thread-safe but needs to be (this instance may be a singleton shared amongst multiple threads).  The runtime will likely include a store barrier on the write to _http3EncodedBytes, but it'd be better to be explicit about it:
```C#
internal byte[] Http3EncodedBytes
{
    get
    {
        byte[]? http3EncodedBytes = Volatile.Read(ref _http3EncodedBytes);
        if (http3EncodedBytes is null)
        {
            http3EncodedBytes = _http3Index is int index && index >= 0 ?
                PackEncoder.EncodeStaticIndexedHeaderFieldToArray(index) :
                QPackEncoder.EncodeLiteralHeaderFieldWithStaticNameReferenceToArray(H3StaticTable.MethodGet, _method);
            Volatile.Write(ref _http3EncodedBytes, http3EncodedBytes);
        }

        return http3EncodedBytes;
    }
}

stephentoub replied:
(Also, the old implementation would be inlined due to being trivial.  The new will not.  I don't know if this property is on any hot paths.)

marek-safar replied:
Why does it need to be thread-safe? Its an internal property which if there is a race would return the same data but in a different instance.

stephentoub replied:
In case the writes to populate the instance were reordered after the write to publish the instance.

marek-safar replied:
Addressed

marek-safar committed:
Merge remote-tracking branch 'upstream/master' into http-cleanup

marek-safar committed:
Remove unreachable code

marek-safar committed:
Add explicit memory barriers

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:
```diff
@@ -499,15 +499,7 @@ private void BufferHeaders(HttpRequestMessage request)
-            }
-            else
-            {
-                BufferLiteralHeaderWithStaticNameReference(H3StaticTable.MethodGet, normalizedMethod.Method);
```
@scalablecory, is this an ok change?

scalablecory replied:
It's potentially less efficient if a customer has already inefficient code and isn't caching custom `HttpMethod` instances. Previously this would never allocate for a custom `HttpMethod`, and now it will.

I'm fine with this change. For the common case it removes a branch, which is probably a worthwhile tradeoff.

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpMethod.cs:
```diff
@@ -19,8 +22,8 @@ public class HttpMethod : IEquatable<HttpMethod>
         private static readonly HttpMethod s_optionsMethod = new HttpMethod("OPTIONS", http3StaticTableIndex: H3StaticTable.MethodOptions);
-        private static readonly HttpMethod s_traceMethod = new HttpMethod("TRACE");
-        private static readonly HttpMethod s_patchMethod = new HttpMethod("PATCH");
+        private static readonly HttpMethod s_traceMethod = new HttpMethod("TRACE", -1);
+        private static readonly HttpMethod s_patchMethod = new HttpMethod("PATCH", -1);
```
I think you meant to make `_http3Index` non-nullable, and to set it to `-1` in the default ctor.

Otherwise, initializing these to `-1` appears to have no meaningful effect.

marek-safar replied:
Initializing to `-1` means that it's a singleton instance with no HTTP3 static
Initializing to `null` means it's not a singleton instance

we could use `-2` instead of null if that's what you are suggesting

scalablecory replied:
Ahhh I see the distinction now. Thanks. Okay, Is good to merge.

scalablecory reviewed (changes requested):
LGTM other than this one comment.
